File: java/parsers/src/main/java/jp/co/teruuu/parser/json/type/JObject.java
package jp.co.teruuu.parser.json.type;

import java.util.Map;

public class JObject {
  Map<JString, JValue> map;

  public JObject(Map<JString, JValue> map) {
    this.map = map;
  }

  @Override
  public boolean equals(Object other) {
    if(!(other instanceof JObject) || this.map.keySet().size() != ((JObject) other).map.keySet().size()) {
      return false;
    }
    for(JString key: map.keySet()) {
      if(!((JObject) other).map.containsKey(key) || !this.map.get(key).equals(((JObject) other).map.get(key))) {
        return false;
      }
    }
    return true;
  }
}

File: scala/parsers/src/main/scala/jp/co/teruuu/parser/json/type/JObject.scala
package jp.co.teruuu.parser.json.`type`

case class JObject(map: Map[JString, JValue]) extends JValue {
  def getMap = map

  override def equals(obj: Any): Boolean = {
    obj match {
      case obj: JObject => {
        (map.size == obj.getMap.size) && map.find(kv => !kv._2.equals(obj.getMap.get(kv._1).get)).isEmpty
      }
      case _ => false
    }
  }
}
